src/main/java/org/opensearch/knn/index/codec/native_index/NativeIndexBuilder.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.codec.native_index;
+
+import java.io.IOException;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.codecs.DocValuesProducer;
+import org.apache.lucene.index.BinaryDocValues;
+import org.apache.lucene.index.FieldInfo;
+import org.opensearch.knn.common.KNNConstants;
+import org.opensearch.knn.index.KNNSettings;
+import org.opensearch.knn.index.VectorDataType;
+import org.opensearch.knn.index.codec.transfer.VectorTransfer;
+import org.opensearch.knn.index.codec.transfer.VectorTransferByte;
+import org.opensearch.knn.index.codec.transfer.VectorTransferFloat;
+import org.opensearch.knn.index.codec.util.KNNCodecUtil;
+import org.opensearch.knn.index.codec.util.SerializationMode;
+import org.opensearch.knn.index.util.KNNEngine;
+import org.opensearch.knn.indices.ModelCache;
+import org.opensearch.knn.plugin.stats.KNNGraphValue;
+
+import lombok.NonNull;
+
+import static org.opensearch.knn.common.KNNConstants.MODEL_ID;
+
+/**
+ * Abstract class to build the KNN index
+ */
+public abstract class NativeIndexBuilder {
```

Comment:
should we think to rename this class to NativeIndexWriter

Reply:
That makes more sense.